### PR TITLE
Convergence tracking

### LINKED ISSUE #149: Docs: Expand on HyperVolume minimum and maximum parameters
Currently the docs and the tutorial don't dive very deep into the [`HyperVolume`](https://github.com/quaquel/EMAworkbench/blob/master/ema_workbench/em_framework/optimization.py#L490) input parameters when used as a convergence metric. In the examples they're (on the surface) quite arbitraily defined, especially the maximums, sometimes being `1`, `1.01` or `1.1`, and sometimes being `2` or `3`. While the docs currently describes what it does and outputs, it doesn't really describe how the input parameters change its behaviour.

The docs and tutorial could include the following things:

- How the input parameters change the HyperVolume behaviour as an convergence metric
- How you can manually (or experimentally) determine realistic minimum and maximum HyperVolume input values
- When and how `HyperVolume.from_outcomes()` can be used (when expected values are defined, etc.)

The docs can be expanded at the [`HyperVolume`](https://github.com/quaquel/EMAworkbench/blob/master/ema_workbench/em_framework/optimization.py#L490) class while the tutorial could be expanded at the [Tracking convergence](https://emaworkbench.readthedocs.io/en/latest/indepth_tutorial/directed-search.html#Tracking-convergence) part of the Directed search tutorial.

### LINKED ISSUE #187: convergence tracking of optimization
It seems there are a few minor issues with convergence tracking in the workbench. First, convergence tracking is triggered before a single run has been executed (which is superfluous). Second, convergence is not tracked one last time after the optimization has reached the stopping condition.

### PR DESCRIPTION
**Pull request by quaquel**
This pull request entails an overhaul of how convergence tracking is being supported by the workbench. This closes #187. 

In short:
1. ArchiveLogger now stores results in a tarbal
2. Additional helper functions for converting from a dataframe to a collection of platypus solutions 
3. Additional helper function for merging results from multiple seeds using epsilon non dominated sort
4. All platypus convergence indicators are wrapped using a wrapper class to make them play nice with the workbench.
5. HyperVolume results in a DeprecationWarning being logged. It is better to calculate hypervolume in post. This also closes #149 
6. modifications to various __init__ files to ensure the correct functions and classes can be directly imported from ema_workbench.

this pull request needs to be merged before we can resolve issue #147

### CODE REVIEW

**Comment by EwoutH** [comment]
Looks like an major effort! I will dive into this tomorrow.

**Review by EwoutH — APPROVED** [review]
Two last nick picks, when those are resolved it's good to go!

**Comment by quaquel** [comment]
> Two last nick picks, when those are resolved it's good to go!


A nice use case for the walrus operator indeed. I had completely. forgetting about it while writing the code.

File: ema_workbench/em_framework/optimization.py [hunk 1]
```diff
@@ -572,25 +808,44 @@ def __init__(self, metrics, max_nfe, convergence_freq=1000, logging_freq=5, log_
...
+        if (nfe >= self.last_check + self.convergence_freq) or (self.last_check == 0) or force:
             self.index.append(nfe)
             self.last_check = nfe
 
             for metric in self.metrics:
                 metric(optimizer)
 
     def to_dataframe(self):
-        progress = {metric.name: metric.results for metric in self.metrics if metric.results}
+        progress = {metric.name: result for metric in self.metrics if result := metric.get_results()}
```
Comment (EwoutH) [inline ema_workbench/em_framework/optimization.py:848]: I forgot what a bitch assignment expressions was [about parentheses](https://peps.python.org/pep-0572/#exceptional-cases). This should fix the syntax error and CI issues.

```suggestion
        progress = {metric.name: result for metric in self.metrics if (result := metric.get_results())}
```

Comment (EwoutH) [reply ema_workbench/em_framework/optimization.py]: @quaquel I will just commit this so we can see if the CI goes green again!

**Comment by EwoutH** [comment]
I checked the Read the Docs and found a few missing empty lines in the docstring, fixed those in 5f8274d, and merged!

Congratulations on getting this major feature in!

**Pull request merged**